feat: handle ipfs messages

il3ven opened the pull request:
This PR adds the ability to handle IPFS messages which are defined in neume-network/schema#40.

The extraction-worker accepts native IPFS URLs like `ipfs://<cid>` and it will be the responsibility of the strategy to format the request as a native URL. For eg. soundxyz will format pinata URLs.  This will also ensure that all the IPFS URLs in music-os are native.

Currently we are not running our own IPFS node therefore a gateway is required in each message.

The extraction-worker also validates CID before sending a request. This will prevent us from sending unnecessary requests and purge bad IPFS URLs. In future, we can also do #45.

This PR also removes @neume-network/message-schema package as that has been deprecated.

Note: In order for tests to pass in GitHub actions we need to merge neume-network/schema#40. Locally all the tests pass.

il3ven committed:
handle messages of `type: "ipfs"`

il3ven committed:
validate CID before making a request

This commit also removes @neume-network/message-schema because it has
been deprecated.

Fixes #33

TimDaub reviewed (changes requested):
- [x] fix unit tests
- [x] rebase against main branch

TimDaub commented on src/api.mjs:
```diff
@@ -162,14 +154,54 @@ async function route(message) {
+  } else if (type === "ipfs") {
+    let { url, gateway, timeout: timeoutFromMsg } = message.options;
+
+    const nativeIPFSPattern = /^(ipfs:\/\/)([^/?#]+)(.*)/;
```
I believe this regex could be part of the schema package or not?

il3ven replied:
This can be part of the schema package but we will need it here too. Because we are using it to split the IPFS message into protocol, hash and CID.

TimDaub replied:
potentiall we could also use a fork if is-ipfs that spits out the CID

TimDaub commented on src/api.mjs:
```diff
@@ -162,14 +154,54 @@ async function route(message) {
+
+    let data;
+    try {
+      data = await request(url, "GET", null, null, signal);
```
a more self-documenting way of writing this is by defining `const body = null` etc and then passing them into the request function

il3ven committed:
upgrade @neume-network/schema package

As we have deprecated message-schema package. The `workerMessage` object
is now exported from the schema package.

il3ven committed:
fix a bug in api_tests

The CI will still not pass because we need to merge
https://github.com/neume-network/schema/pull/40 first.

il3ven committed:
improve readability of code

il3ven commented:
@TimDaub The unit tests will pass after neume-network/schema#40 has been merged. I have tested the PR locally using npm link and the tests do pass.

We can also see in the [latest logs](https://github.com/neume-network/extraction-worker/actions/runs/3185292874/jobs/5194696426#step:6:219) that only the tests related to IPFS fail because they can't validate the message.

TimDaub commented on package.json:
```diff
@@ -59,16 +59,16 @@
   "dependencies": {
-    "@neume-network/message-schema": "0.3.1",
-    "@neume-network/schema": "0.5.0",
+    "@neume-network/schema": "^0.6.0",
```
```suggestion
    "@neume-network/schema": "0.6.1",
```

let's not use the caret as it is dangerous and can lead to supply chain attacks

TimDaub committed:
Update package.json

TimDaub commented:
we need to update package-lock.json to have the unit tests pass

TimDaub committed:
Update package-lock.json

il3ven committed:
change url to uri in ipfs messages

This change was prompted by an update in neume-network/schema.

il3ven commented:
@TimDaub This PR is now ready to be merged. Just needed to change `url` to `uri` because of change in schema.

TimDaub commented:
awesome thanks
